lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -545,84 +546,203 @@ def as_dict_with_promql_labels(self):
+    #   the "alert" and "expr" keys.
+    # - alert rule (singular): a single dictionary that has the "alert" and "expr" keys.
 
-            # insert juju topology filters into a prometheus alert rule
-            rule["expr"] = topology.render(rule["expr"])
-            return rule
+    def __init__(self, topology: JujuTopology):
+        self.topology = topology
+        self.alert_groups = []  # type: List[dict]
 
+    @staticmethod
+    def _is_valid_official_format(rules_dict: dict) -> bool:
+        # TODO validate against schema using e.g. pydantic
+        return "groups" in rules_dict
 
-def load_alert_rules_from_dir(
-    dir_path: Union[str, Path], topology: JujuTopology, *, recursive: bool = False
-) -> List[dict]:
-    """Load alert rules from rule files.
+    @staticmethod
+    def _is_valid_lma_format(rules_dict: dict) -> bool:
+        # one alert rule per file
+        # TODO validate against schema using e.g. pydantic
+        return set(rules_dict) >= {"alert", "expr"}
 
-    All rules from files for the same directory are loaded into a single
-    group. The generated name of this group includes juju topology.
-    By default, only the top directory is scanned; for nested scanning, pass `recursive=True`.
+    @staticmethod
+    def _from_file(root_path: str, file_path: Path, topology: JujuTopology) -> List[dict]:
+        """Read a rules file from path, injecting juju topology.
 
-    Args:
-        dir_path: directory containing *.rule files (alert rules without groups).
-        topology: a `JujuTopology` instance.
-        recursive: flag indicating whether to scan for rule files recursively.
+        Args:
+            root_path: path to the root rules folder (used only for generating group name)
+            file_path: path to a *.rule file.
+            topology: juju topology information.
 
-    Returns:
-        a list of prometheus alert rule groups.
-    """
-    alerts = defaultdict(list)
+        Returns:
+            A list of dictionaries representing the rules file, if file is valid (the structure is
+            formed by `yaml.safe_load` of the file); an empty list otherwise.
+        """
+        with file_path.open() as rf:
+            # Load a list of rules from file then add labels and filters
+            try:
+                rule_file = yaml.safe_load(rf)
+
+            except Exception as e:
+                logger.error("Failed to read alert rules from %s: %s", file_path.name, e)
+                return []
+
+            # TODO validate input
+            if AlertRules._is_valid_official_format(rule_file):
+                alert_groups = rule_file["groups"]
+            elif AlertRules._is_valid_lma_format(rule_file):
+                # convert to list of alert groups
+                # group name is made up from the file name
+                alert_groups = [{"name": file_path.stem, "rules": [rule_file]}]
```

Comment:
Am I correct in understanding this means for rules in LMA format each rule is put into a group of its own ? If so is this as per current product specification ? 
Note currently all rules in LMA format are put into a common group.

Reply:
Correct, for the custom format ("LMA format"), group name is made up from the filename, per (current) spec.